distributed/protocol.py

```diff
@@ -64,35 +62,77 @@
     default_compression = 'lz4'
 
 
+def _extract_big_bytes(x, big, path=()):
+    if type(x) is dict:
+        for k, v in x.items():
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                del x[k]
+    elif type(x) is list:
+        for k, v in enumerate(x):
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                x[k] = None
+
+
 def dumps(msg):
```

Comment:
I find no noticeable change between this PR and master when running that benchmark.

Trying all my usual python micro-optimization tricks, I can't get a noticeable speedup on the intentionally bad case. I think the only way to speed this up (if needed) would be to do less work. One option would be to set a limit on the depth at which we traverse, another would be to add a boolean (as you suggested above) and use full traversal only where necessary. We could also only traverse certain keys.